### PR TITLE
feat(cloudrunner): recover panic

### PR DESCRIPTION
**Pull request by philiphassel**
Recover panic from service and log as error before termination.

### CODE REVIEW

File: run.go [hunk 1]
```diff
@@ -131,6 +131,15 @@ func Run(fn func(context.Context) error, options ...Option) error {
 		zap.Object("buildInfo", buildInfoMarshaler{buildInfo: buildInfo}),
 	)
 	defer logger.Info("goodbye")
+	defer func() {
+		if r := recover(); r != nil {
+			logger.Error(
```
Comment (odsod) [inline run.go:136]: do we need to have the error return value as a named argument and assign to it here in order for the function to still exit with an error?

Comment (philiphassel) [reply run.go:136]: Yes. We could do it, I'll fix.

File: run.go [hunk 2]
```diff
@@ -131,6 +131,15 @@ func Run(fn func(context.Context) error, options ...Option) error {
...
+				"recovered panic",
+				zap.Any("msg", r),
```
[the remark this answers is not in the document]
Comment (philiphassel) [reply run.go:138]: Sure!

File: run.go [hunk 3]
```diff
@@ -131,6 +131,23 @@ func Run(fn func(context.Context) error, options ...Option) error {
...
 	defer logger.Info("goodbye")
+	defer func() {
+		if r := recover(); r != nil {
+			var msg zap.Field
+			if err2, ok := r.(error); ok {
+				msg = zap.Error(err2)
+				err = err2
+			} else {
+				msg = zap.Any("msg", r)
+				err = fmt.Errorf("recovered panic")
```
Comment (philiphassel) [inline run.go:142]: I think a generic error message like this is fine since all the other information will be logged as error. What do you think @odsod ?

**Review by odsod — APPROVED** [review]

**Review by thall — APPROVED** [review]

**Pull request merged**